### PR TITLE
refactor: markdown

### PR DESCRIPTION
**Pull request by floriandwt**
This PR introduces a new markdown rendering enabling more interop. between e.g. GitHub Readmes, Marketplace documentations, and can be used in further use cases. See https://github.com/inlang/monorepo/issues/1302

### CODE REVIEW

**Comment by samuelstroschein** [comment]
@floriandwt can you copy the pacakge-lock.json fom the main branch, delete the package-lock.json in this branch, and run `npm run ci` afterwards? 

the lock file got overwritten or something leading to "15.000 line of change"

File: inlang/documentation/reference/tableOfContents.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+[
+	{
+		"path": "path-to-file.md",
+		"slug": "slug-of-url", // has to be unique
+		"title": "Title of the page",
+		"description": "Description of the page",
+	}
+]
+```
+
```
Comment (samuelstroschein) [inline inlang/documentation/reference/tableOfContents.md:21]: @floriandwt 

left over from the table of contents on marketplace?

this seems to be redundant. we don't need to expose internal stuff on the docuementaiton site.

File: inlang/source-code/markdown/package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		"./src",
+		"./dist"
+	],
+	"exports": {
+		".": "./dist/index.js",
+		"./custom-elements": "./dist/custom-elements/index.js",
+		"./css": "./dist/css/index.js"
+	},
+	"scripts": {
+		"build": "npx playwright install --with-deps chromium && npx tailwindcss -i ./src/css/style.css -o ./dist/css/style.css && npx prettier ./src --write && tsc --build",
```
Comment (samuelstroschein) [inline inlang/source-code/markdown/package.json:16]: @floriandwt I am unsure if we should support markdown: 

- hundreds of dependencies
- requirement for a headless browser which slows down testing, ci/cd, install scripts, and more. 

Let's take out mermaid for now, and remove the unproportional amount of complexity entailed by mermaid support. 

I am looking at https://d2lang.com/ as a replacement because d2 can render without a headless browser, and has a nicer syntax. Unfortunately, no JS bindings exists yet.

File: inlang/source-code/website/src/pages/blog/@id/index.page.tsx [hunk 3]
```diff
@@ -1,31 +1,112 @@
...
+		}
+		window.history.pushState({}, "", `${currentPageContext.urlParsed.pathname}#${anchor}`)
+	}
+
+	onMount(async () => {
+		for (const heading of props.markdown
+			.match(/<h[1-3].*?>(.*?)<\/h[1-3]>/g)
+			.map((heading: string) => {
+				// We have to use DOMParser to parse the heading string to a HTML element
+				const parser = new DOMParser()
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/pages/blog/@id/index.page.tsx:58]: What is all this code for? Why the replacements?

File: inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx [hunk 4]
```diff
@@ -1,80 +1,57 @@
...
+		const markdown = await convert(raw)
 
-		index[markdown.frontmatter.href] = markdown
+		index[content.slug] = markdown
+		generatedTableOfContents[content.slug] = {
+			...content,
+			href: "/blog/" + content.slug,
+		}
 	}
 }
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx:57]: You don't need to generate the table of contents anymore. We already have it in the form of the JSON. 

**What to change**

1. create a variable `renderedMarkdown` at the root level which will store already rendered markdown.
2. In `onBeforeRender()`, check if the requested blog is contained in `renderedMarkdown`. Otherwise, render it and store the result in `renderedMarkdown`. 
3. Remove the `generateIndexAndTableOfContents()` function. 
4. Remove `processedTableOfContents` from the page context as the client can directly import the json
5. REPEAT IN DOCUMENTATION SITE

```ts
const renderedMarkdown = {}

export async function onBeforeRender(pageContext){
  const slug = pageContext.urlPathname.slice(`/blog/`.length)
  if (renderedMarkdown[slug] === undefined){
    // pseudocode
    const content = tableOfContents.find(slug)
    const text = await fs.readFile(new URL(`blog/${content.path}`, repositoryRoot), "utf-8")
    const markdown = await convert(text)
    renderedMarkdown[slug] = markdown
  }
  return {
    pageContext: {
      pageProps: {
        markdown: renderedMarkdown[slug]
      }
    }
  }
}
```

File: inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx [hunk 5]
```diff
@@ -1,80 +1,57 @@
...
 		throw RenderErrorPage({ pageContext: { is404: true } })
 	}
+
 	return {
 		pageContext: {
 			pageProps: {
-				markdown: index[pageContext.urlPathname]!,
-				processedTableOfContents: processedTableOfContents,
+				markdown: index[pageContext.urlPathname.replace("/blog/", "")]!,
+				meta: generatedTableOfContents[pageContext.urlPathname.replace("/blog/", "")]!,
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx:39]: What is meta for?

File: inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx (hunk 4 above)
Comment (samuelstroschein) [reply inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx:57]: The code above is all we need to render markdown :)

File: inlang/source-code/website/package.json [hunk 6]
```diff
@@ -53,7 +51,6 @@
 		"flat": "^5.0.2",
 		"isomorphic-git": "1.24.5",
 		"jose": "^4.11.1",
-		"markdown-to-txt": "^2.0.1",
 		"marked": "^4.2.4",
 		"mermaid": "^9.3.0",
```
Comment (samuelstroschein) [inline inlang/source-code/website/package.json:55]: Those two dependencies seem markdown service related. Try to remove them and see if everything still works.

File: inlang/source-code/website/package.json [hunk 7]
```diff
@@ -89,6 +88,7 @@
 		"fast-glob": "^3.2.12",
 		"patch-package": "8.0.0",
 		"postcss": "^8.4.18",
+		"rehype-accessible-emojis": "0.3.2",
```
Comment (samuelstroschein) [inline inlang/source-code/website/package.json:91]: Wrong dependency? Why does the website need to install a rehype dependency?

File: inlang/source-code/website/package.json [hunk 8]
```diff
@@ -101,5 +101,10 @@
 	"peerDependencies": {
 		"express": "^4.18.2"
 	},
-	"license": "Apache-2.0"
+	"license": "Apache-2.0",
+	"description": "This module is a monolith that contains inlang's website and editor to manage translations.",
+	"version": "1.0.0",
+	"main": "index.js",
+	"keywords": [],
+	"author": ""
```
Comment (samuelstroschein) [inline inlang/source-code/website/package.json:109]: 1. Why did you add those properties? 
2. Remove them. This pacakge is not published to NPM. Main doesn;'t even exist.

**Review by samuelstroschein — CHANGES_REQUESTED** [review]
Overall as expected. One remark: 

The mermaid feature seems to introduce a lot (too much) complexity. I assume that felt at one point "hm, adding a playwright dependency doesn't seem right". Speak up and discuss with your team members. Your gut feeling is usually right. The engineering goal is not to support any feature but to support a feature that is _worth it given the compexity it introduces_. Otherwise, the project will suffer from unreasonable long term maintainence burden(s).

File: inlang/source-code/website/src/pages/blog/@id/index.page.tsx (hunk 3 above)
Comment (floriandwt) [reply inlang/source-code/website/src/pages/blog/@id/index.page.tsx:58]: This is the code of the tableOfContents and the code of scrolling to the specific anchors (at it is not working otherwise, -> seen yesterday). Also replace function necessary to sanitize the headings from emojis which others have been used in the readmes.

File: inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx (hunk 5 above)
Comment (floriandwt) [reply inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx:39]: SEO

Comment (samuelstroschein) [reply inlang/source-code/website/src/pages/blog/@id/index.page.server.tsx:39]: But you get that from the description don't you? If so, name it description :)

**Comment by floriandwt** [comment]
@samuelstroschein almost all remaining should be solved now, I will finish getting rid of mermaid on Monday and move the new tableOfContents function to the client then.
